feat(config): Mongoose 4.11 upgrade

mleanos opened the pull request:
Upgrades Mongoose to ~~4.11.1~~ 4.11.3

Updates Mongoose connection implementation to accommodate deprecated
features & connection options.

Also, updates the Gulp Mocha tasks to reflect changes to the Mongoose
implementation.

Fixes tests to get the database from the existing Mongoose singleton.

Derives from changes in https://github.com/meanjs/mean/pull/1816

Closes https://github.com/meanjs/mean/issues/1814

mleanos commented:
I started on a bigger refactor of the Mongoose and Express app implementation. However, I scaled it back to just satisfy the Mongoose upgrade. We can discuss a bigger refactor in a separate issue and/or PR.

Regarding the `{ user: '', pass: '' }` connection options, there is [work being done](https://github.com/mongodb/node-mongodb-native/pull/1529) to add back support of this feature.

lirantal commented:
@mleanos are you synced with @simison? is this a superset of his PR?

simison commented on config/lib/mongoose.js:
```diff
@@ -24,25 +25,28 @@ module.exports.connect = function (cb) {
-  });
+      if (cb) cb(connection.db);
+    })
+    .catch((err) => {
```
some ES6 here ;-)

mleanos replied:
Do you think it's ok?

Node.js `6.4.0>` can handle handle arrow functions. So why not?

simison replied:
I'd love to see bigger discussion about this, actually. This topic just came up at Trustroots so I might open an issue here just to get some food for thought.

I don't think it's a good idea to introduce mixed coding style in PR like this.

It's more annoying to me to see mixed style rather than old school JS.

I'd love to see meanjs move on to ES6 but it might be better done with some strategy and thought and build tools (eslint) could probably be upgraded first.

WDYT?

mleanos replied:
Everything you just said seems very reasonable to me. I'd like to discuss the introduction of ES6 as well. I think it started becoming a habit for me, since I'm using some features like array functions in my projects.

I'll remove my ES6 usage from this PR.

simison commented on gulpfile.js:
```diff
@@ -296,10 +296,7 @@ gulp.task('mocha', function (done) {
-        mongoose.disconnect(function () {
-          done(error);
-        });
+        mongooseService.disconnect(done);
```
I almost did this myself as well, but that `error` being passed to gulp's `done` isn't error from mongoose disconnect. It's the error object from Mocha.

Perhaps that variable needs renaming so that it would be clear what's happening here?

mleanos replied:
Yea, good points. It did occur to me, for a moment, about the `error` needing to be passed to the done. Then I think I just overlooked it. I'll change it to pass the Mocha error to the `done` callback. I guess we can just log the error to the console, when the `disconnect` fails; it would be a rare error in this case.

The name `error` seems fine to me. It's clear what's going on since it's right there.

simison commented on gulpfile.js:
```diff
@@ -362,16 +359,17 @@ gulp.task('karma:coverage', function(done) {
 gulp.task('dropdb', function (done) {
   // Use mongoose configuration
-  var mongoose = require('./config/lib/mongoose.js');
+  var mongooseService = require('./config/lib/mongoose');
```
Same as earlier:

Is there any reason why this `require` is inside the task and not on top of the gulp file with all the other dependencies?

mleanos replied:
If our Mongoose service is only being used in a couple places, why `require` it at the global scope?

My understanding of how `require` works, along with it's memory use, isn't great so maybe it doesn't matter.

simison replied:
Then I don't know more than you do. :-) I was just curious, I guess for now we can trust someone knew what they were doing when they placed it inside the task.

lirantal replied:
Amos added it way back in 0.4 release branch.
We can move it at the top if it makes sense now, but I think when Amos added it on 0.4 the mongoose.js worked differently from today, as in, just requiring it may have opened the database connection or instantiate models so that would explain why he would have wanted to limit it to the specific gulp tasks.

simison commented on gulpfile.js:
```diff
@@ -277,15 +277,15 @@ gulp.task('templatecache', function () {
 gulp.task('mocha', function (done) {
-  // Open mongoose connections
-  var mongoose = require('./config/lib/mongoose.js');
+  var mongooseService = require('./config/lib/mongoose');
```
Is there any reason why this `require` is inside the task and not on top of the gulp file with all the other dependencies?

simison commented on modules/articles/tests/server/admin.article.server.routes.tests.js:
```diff
@@ -23,7 +23,7 @@ var app,
   before(function (done) {
     // Get application
-    app = express.init(mongoose);
+    app = express.init(mongoose.connection.db);
```
This <3, I just couldn't figure out what was wrong.

mleanos replied:
This was a source of confusion to me too. I didn't see how the line above would work by just passing `mongoose`, but eventually it made sense based on this: https://github.com/meanjs/mean/blob/master/config/lib/mongoose.js#L27. `db` is a promise, no?

If these changes get merged then it's a moot point, I guess. Ideally, we need to refactor the Express & Mongoose implementations. As it stands, after we make the first connection, we're heavily relying on the fact that Mongoose is a singleton. This could be confusing, and probably not the best SOC design.

simison replied:
Yup! 

I guess https://github.com/lirantal/Riess.js might have quite a lot of it thought already.

simison replied:
And that said I think it's a good idea to keep this to minimum to just get Mongoose upgraded and leave refactoring to a separate PR.

mleanos replied:
Agreed 👍

lirantal replied:
Yep. And indeed the returned value of the mongoose.js line 27 was just the db object being resolved with.

mleanos commented:
@lirantal Mikael and I discussed in his other PR that I would be taking this over. Once I get a couple things clarified from the review comments, I'll rebase and make the necessary changes.

mleanos commented:
@lirantal @simison Final review?

Based on Mikael's review:
1) Removed arrow function use
2) Added back `done(error)` to the Mocha Gulp task to return error thrown by Mocha rather than the `mongooseService.disconnect` callback.

simison commented:
Looks good to me now!

mleanos commented on config/env/development.js:
```diff
@@ -5,10 +5,7 @@ var defaultEnvConfig = require('./default');
-      user: '',
-      pass: ''
-    },
+    options: {},
```
Sure thing.

mleanos replied:
Done

mleanos commented:
@simison @lirantal I've rebased and squashed my commits.

I also removed the `options` from https://github.com/meanjs/mean/blob/master/config/env/default.js#L10, that I had previously added in a commit on this branch. I just remembered that the default env config doesn't need the `options`, since they are being set in the individual env configs (i.e. development, production, and test).

lirantal commented on modules/articles/tests/server/article.server.routes.tests.js:
```diff
@@ -119,7 +119,7 @@ describe('Article CRUD tests', function () {
     articleObj.save(function () {
       // Request articles
-      request(app).get('/api/articles')
+      agent.get('/api/articles')
```
nice catch @mleanos

lirantal commented on modules/users/server/config/users.server.config.js:
```diff
@@ -11,7 +11,7 @@ var passport = require('passport'),
  * Module init function
  */
-module.exports = function (app, db) {
+module.exports = function (app) {
```
this could potentially break vertical modules that rely on it.
even with the changes in this PR for `config/lib/mongoose.js` how would one be able to access the db instance?

the other alternative I can think of is if we set app.local.db or something like that which would expose the db connection where app is too.

WDYT?

simison replied:
I like that, sounds simple and straightforward. I didn't like at all trying to follow where `db` is going around trough all those callbacks.

That said, this could also wait for more thorough restructuring of mongoose/db stuff as with promises etc.

lirantal replied:
In leaning towards agreeing with 'this can wait' but I wonder what projects built based on master version will do if `db` is now not available for them and they wish to stay up to date...

mleanos replied:
We should note the potential breaking change in the CHANGELOG. 

Anyone merging from master should be aware of the changes, and it seems like it would be easy for them to add back the `db` in their exports where they rely on it. Adding `app.local.db` wouldn't be a difficult task either.

I'm a bit skeptical that many use cases would arise that would require someone to rely on accessing the `db` instance from server-side config/controllers.

I'm not opposed to adding `db` back to all the callbacks though.

lirantal replied:
yeah which is why I'm not totally against removing it.
so let's keep this PR as is and see how the mongodb refactoring continues on. I'm ok with that.

mleanos replied:
Ok. I'm fine with that as well.

mleanos commented:
@lirantal Are you still waiting on changes before you can approve this?

lirantal commented:
Nope, I'm good.

simison commented on config/lib/mongoose.js:
```diff
@@ -24,25 +25,28 @@ module.exports.connect = function (cb) {
 
```
Remove `var _this = this;` as it's never used (weird how eslint didn't warn about this?)

mleanos replied:
Removed. The ESLint rule [no-unused-vars is turned off](https://github.com/meanjs/mean/blob/master/.eslintrc.js#L36), probably because there are a lot of ESLint errors generated from this rule in our tests and other places :/ That can be addressed in a separate PR.

simison commented on config/lib/mongoose.js:
```diff
@@ -24,25 +25,28 @@ module.exports.connect = function (cb) {
-      if (cb) cb(db);
-    }
-  });
+      if (cb) cb(connection.db);
```
Personally I prefer using `callback` instead of `cb` just to keep it all the same and I dislike abbreviations in coding (just needs extra thinking when reading it), but this isn't a biggie. :-)

mleanos commented:
@simison I addressed the changes requested in your last review. LGTY?

@lirantal If everything looks good now, I think this is ready to be merged.

lirantal commented:
Yep, I'm good.

simison commented:
LGTM! Feel free to merge.
